File: old/Privault.Code/AppVersion.cs
using System.Reflection;

namespace Privault.Core;

/// <summary>
/// Stellt Versionsinformationen der Anwendung bereit.
/// </summary>
/// <remarks>
/// Diese Klasse ermöglicht den Zugriff auf die Versionsnummer der Anwendung (Major, Minor, Patch)
/// sowie auf die minimal erforderliche Server-Version. Die Versionsinformationen werden aus den
/// Assembly-Metadaten extrahiert und lazy-initialisiert.
/// </remarks>
public static class AppVersion
{
    // ------------------------------------------------------------------------
    // --- Eigenschaften ---
    // ------------------------------------------------------------------------

    /// <summary>
    /// Major-Version der App.
    /// </summary>
    /// <remarks>
    /// Wird nach einem Redesign oder bei einem Migrations-Bruch erhöht.
    /// </remarks>
    public static int Major => Version.Value.Major;

    /// <summary>
    /// Minor-Version der App.
    /// </summary>
    /// <remarks>
    /// Wird nach Änderung der Funktionalität erhöht. Wird auf 0 zurückgesetzt, wenn MAJOR erhöht wird.
    /// </remarks>
    public static int Minor => Version.Value.Minor;

    /// <summary>
    /// Patch-Version der App.
    /// </summary>
    /// <remarks>
    /// Wird nach einer Fehlerbehebung (Bugfix) erhöht. Wird auf 0 zurückgesetzt, wenn MAJOR oder MINOR erhöht wird.
    /// </remarks>
    public static int Patch => Version.Value.Build >= 0 ? Version.Value.Build : 0;

    /// <summary>
    /// Minimal erforderliche Server-Minor-Version.
    /// </summary>
    public static int RequiredServerMinor => RequiredServerMinorAttribute.Value;
    
    // ------------------------------------------------------------------------
    // --- Private Eigenschaften ---
    // ------------------------------------------------------------------------

    /// <summary>
    /// Ermöglicht den Zugriff auf die Assembly, die die Kernanwendungslogik enthält.
    /// </summary>
    /// <remarks>
    /// Diese Eigenschaft verweist auf die Assembly, in der die Klasse <c>AppVersion</c> definiert ist.
    /// Sie wird verwendet, um Metadaten auf Assembly-Ebene zu extrahieren, wie z.B. Versionsinformationen und benutzerdefinierte Attribute.
    /// </remarks>
    private static Assembly CoreAssembly => typeof(AppVersion).Assembly;
    
    /// <summary>
    /// Liest die System-Version aus den Assembly-Metadaten und bereinigt sie.
    /// </summary>
    /// <remarks>
    /// Lazy-initialisierte Versionsinformation der Anwendung, extrahiert aus dem <see cref="AssemblyInformationalVersionAttribute"/>.
    /// </remarks>
    private static readonly Lazy<Version> Version = new(() =>
    {
        var info = CoreAssembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion ?? "0.0.0";

        // SemVer: "1.2.3-rc.1+meta" -> "1.2.3"
        var clean = info.Split('-', '+')[0];
        
        // Robust: kein Throw bei kaputter Version, sondern Fallback
        return System.Version.TryParse(clean, out var parsed) ? parsed : new Version(0, 0, 0);
    });
    
    /// <summary>
    /// Liest die minimal erforderliche Server-Minor-Version aus den Assembly-Metadaten.
    /// </summary>
    /// <remarks>
    /// Lazy-initialisierter Wert, der aus dem <see cref="AssemblyMetadataAttribute"/> mit dem Schlüssel "RequiredServerMinor" extrahiert wird.
    /// </remarks>
    private static readonly Lazy<int> RequiredServerMinorAttribute = new(() =>
    {
        var attribute = CoreAssembly.GetCustomAttributes<AssemblyMetadataAttribute>()
            .FirstOrDefault(a => a.Key == "RequiredServerMinor");
        var value = attribute?.Value;
        return value != null && int.TryParse(value, out var parsed) ? parsed : 0;
    });
}

File: lib/core/app_version.dart
import 'package:package_info_plus/package_info_plus.dart';
import 'package:famkey/core/sync_protocol.dart';
import 'package:famkey/database/database.dart';

/// Zentrale Versionierungsinformationen.
///
/// Alle Werte stammen aus ihren jeweiligen Single Sources of Truth.
class AppVersion {
  const AppVersion._(); // verhindert Instanziierung

  /// Angezeigte App-Version (aus `pubspec.yaml`, z.B. "1.0").
  static Future<String> get version async {
    final info = await PackageInfo.fromPlatform();
    return info.version;
  }

  /// Hauptversion der App (aus `version` extrahiert).
  /// Wird mit einem Redesign oder bei einem Migrations-Bruch erhöht.
  static Future<int> get major async {
    final info = await PackageInfo.fromPlatform();
    final parts = info.version.split('.');
    return int.tryParse(parts[0]) ?? 0;
  }

  /// Nebenversion der App (aus `version` extrahiert).
  /// Wird mit einer Funktionsänderung erhöht und mit einer neuen Hauptversion auf 0 zurückgesetzt.
  static Future<int> get minor async {
    final info = await PackageInfo.fromPlatform();
    final parts = info.version.split('.');
    return parts.length > 1 ? int.tryParse(parts[1]) ?? 0 : 0;
  }

  /// Patch der Nebenversion (aus `version` extrahiert).
  /// Wird mit einer Fehlerbehebung (Bugfix) erhöht und mit einer neuen Nebenversion auf 0 zurückgesetzt.
  static Future<int> get patch async {
    final info = await PackageInfo.fromPlatform();
    final parts = info.version.split('.');
    return parts.length > 2 ? int.tryParse(parts[2]) ?? 0 : 0;
  }

  /// Buildnummer (aus `pubspec.yaml`, z.B. 42).
  /// Wird (theoretisch) mit jedem Build erhöht. Sie wird niemals zurückgesetzt.
  /// Dies ist auch der `versionCode` für den Google-Store.
  static Future<int> get buildNumber async {
    final info = await PackageInfo.fromPlatform();
    return int.tryParse(info.buildNumber) ?? 0;
  }

  /// Angezeigte App-Version inkl. Buildnummer (aus `pubspec.yaml`, z.B. "1.0+42")
  static Future<String> get fullVersion async {
    final info = await PackageInfo.fromPlatform();
    return "${info.version}+${info.buildNumber}";
  }

  /// Sync-Protokollversion (aus `lib/core/sync_protocol.dart`)
  static int get syncProtocolVersion => SyncProtocol.version;

  /// Datenbankschema (aus `lib/database/database.dart`).
  static int get databaseSchemaVersion => AppDatabase.version;
}
